db/data_listeners.cc

```diff
@@ -80,25 +80,46 @@ future<> toppartitions_data_listener::stop() {
 
 flat_mutation_reader toppartitions_data_listener::on_read(const schema_ptr& s, const dht::partition_range& range,
         const query::partition_slice& slice, flat_mutation_reader&& rd) {
-    if (s->ks_name() != _ks || s->cf_name() != _cf) {
-        return std::move(rd);
+    if (!_families) {
+        dblog.trace("toppartitions_data_listener::on_read: {}.{}", s->ks_name(), s->cf_name());
+        return make_filtering_reader(std::move(rd), [zis = this->weak_from_this(), &range, &slice, s = std::move(s)] (const dht::decorated_key& dk) {
+            // The data query may be executing after the toppartitions_data_listener object has been removed, so check
+            if (zis) {
+                zis->_top_k_read.append(toppartitions_item_key{s, dk});
+            }
+            return true;
+        });
     }
-    dblog.trace("toppartitions_data_listener::on_read: {}.{}", s->ks_name(), s->cf_name());
-    return make_filtering_reader(std::move(rd), [zis = this->weak_from_this(), &range, &slice, s = std::move(s)] (const dht::decorated_key& dk) {
-        // The data query may be executing after the toppartitions_data_listener object has been removed, so check
-        if (zis) {
-            zis->_top_k_read.append(toppartitions_item_key{s, dk});
+
+    for (const auto& [ks, cf] : _families.value()) {
```

Comment:
I guess we don't expect it to be too big, because you'll do this loop for any operation and you don't have that quick bale-out in the original code where you compare the ks/table and return imidiately

Reply:
It shouldn't be big, but I'll change it to a map as you pointed out. It's a good point.